http/analysis-service/server.py

```diff
@@ -846,6 +859,146 @@ def kill(instance_id):
         monitoring_url  = abs_url_for('monitor', instance_id = instance.id)
     )
 
+@app.route("/cluster", methods=["GET"])
+@login_required
+def cluster_get_params(errors=None, values=None):
+    # Check that the user logged in is also authorized to do this
+    if not current_user.is_authorized():
+        return login_manager.unauthorized()
+    return render_template('cluster/cluster.html', errors=errors, values=values,
+        token=str(uuid4()))
+
+@app.route("/cluster/new", methods=["POST"])
+@login_required
+def cluster_spawn():
+    # Check that the user logged in is also authorized to do this
+    if not current_user.is_authorized():
+        return login_manager.unauthorized()
+
+    errors = {}
+
+    # Check required fields
+    for f in ['name', 'token', 'n-workers']:
+        val = request.form[f]
+        if val is None or val.strip() == '':
+            errors[f] = "This field is required"
+
+    try:
+        n_workers = int(request.form["n-workers"])
+        if n_workers <= 0:
+            raise Exception
+    except:
+        errors["n-workers"] = "This field should be a positive number"
+
+    # Check required file
+    if not request.files['public-ssh-key']:
+        errors['code-tarball'] = "Public key file is required"
+
+    # Bug 961200: Check that a proper OpenSSH public key was uploaded.
+    # It should start with "ssh-rsa AAAAB3"
+    pubkey = request.files['public-ssh-key'].read()
+    if not validate_public_key(pubkey):
+        errors['public-ssh-key'] = "Supplied file does not appear to be a valid OpenSSH public key."
+
+    if errors:
+        return cluster_get_params(errors, request.form)
+
+    # Create EMR cluster
+    n_instances = n_workers if n_workers == 1 else n_workers + 1
+
+    install_spark_bootstrap = BootstrapAction('Install Spark',
+                                              's3://support.elasticmapreduce/spark/install-spark',
+                                              ['-v', app.config['SPARK_VERSION']])
+
+    setup_telemetry_bootstrap = BootstrapAction('Setup Telemetry',
+                                                's3://telemetry-spark-emr/telemetry.sh',
+                                                ['--public-key', pubkey])
+
+    jobflow_id = emr.run_jobflow(name = request.form['token'],
+                              ec2_keyname = 'mozilla_vitillo',
```

Comment:
We should eventually not use our own keys, but I see I already did the same thing for the worker instances, so let's not worry about it for now.